Meme/Meme/MemeTemplatesViewController.swift

```diff
@@ -38,6 +54,33 @@ class MemeTemplatesViewController: UIViewController {
     }
 }
 
+extension MemeTemplatesViewController : UITableViewDelegate, UITableViewDataSource {
+
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return self.templates.count
+    }
+
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = self.tableView.dequeueReusableCell(withIdentifier: cellID) as! MemeTemplatesTableViewCell
+        cell.memeImageView.image = nil
+
+        if let imageURL = self.templates[indexPath.row].imageURL {
+            if let cachedImage = self.imageCache[imageURL.absoluteString] {
+                cell.memeImageView.image = cachedImage
+            } else {
+                cell.memeImageView.load(url: imageURL) { image in
+                    self.imageCache[imageURL.absoluteString] = image
```

Comment:
I think the load method should handle cache retrieval

Reply:
Hmm, the load method is defined as an extension on UIView though, and extensions can't contain stored properties 🤔

Reply:
The image cache should be its own object. you could make it a singleton but wrap it up as a service
```swift
protocol ImageCaching {
    getImage(for url: URL) -> UIImage?
    setImage(for url: URL)
}
```
```swift
class ImageCache: ImageCaching {
    static let shared = ImageCache()

    private var imageDictionary: [URL: UIImage] = [:]

    ...
}
```

```swift
extension UIView {
    load(url: URL, completion: (UIImage) -> Void, cache: ImageCaching = ImageCache.shared) { ... }
}
```
Something like this?

Reply:
Gotcha nice!

Reply:
https://github.com/vhart/MartianTalk/blob/master/MartianNewsReader/Models/Services/Data%20Retrieval/ImageStore.swift